### PR TITLE
session: add test for local temporary table update

### LINKED ISSUE #26051: Support update for local temporary table
Depends for #25920 for update need output from executors like TableReader

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #26051 

### What is changed and how it works?

Add test for local temporary table update

### Check List 

Tests 

- Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by tiancaiamao — COMMENTED** [review]
Execuse me, are you writing javascript?

File: session/session_test.go [hunk 1]
```diff
@@ -5149,3 +5150,232 @@ func (s *testSessionSuite) TestLocalTemporaryTableScan(c *C) {
...
+		if !expectedUpdateSuccess {
+			checkUpdatesAndDeletes(nil, nil)
+		}
+		verify(err)
+		tk.MustExec("delete from tmp1")
+		tk.MustQuery("select * from tmp1").Check(testkit.Rows())
+	}
+
+	// update with point get for primary key
+	assertUpdate("update tmp1 set v=999 where id=1", true, func(_ error) {
```
Comment (tiancaiamao) [inline session/session_test.go:5263]: Can you make the tests table driven, and use less callback function as possible?

For example, 

```
type TestCase struct {
    sql string
    expected string
}

for _, c := range testCases {
    testCase(c)
}

func testCase() {
    // the assertUpdate function
}
```
`checkUpdatesAndDeletes` do not really need to be a callback function, most of the case do not need that flexible... If you really want that, you can make it data oriented, for example, 

```
 testCases := []testCase{
    {"update tmp1 set id=12 where id=1",  nil},
    {"update tmp1 set id=12 where id=1", checkResult{12, 101, 1001}},
    {"update tmp1 set id=1 where id=1", checkResult{}},
    {},
}


type checkResult  struct {
   ...
}

func (checkResult) Check() {
  .....
}
```

In this way, the test cases are driven by a table, and if you think the pure data is not flexible enough, you can make them into data struct and wrap method on them (as the callback function) ...

File: session/session_test.go [hunk 2]
```diff
@@ -5149,3 +5150,185 @@ func (s *testSessionSuite) TestLocalTemporaryTableScan(c *C) {
...
+		for _, id := range otherIds {
+			modify, exist := modifyMap[id]
+			c.Assert(exist, IsTrue)
+			expect = append(expect, modify)
+		}
+
+		tk.MustQuery("select * from tmp1").Check(testkit.Rows(expect...))
+	}
+
+	executeSql := func(sql string, checkResult interface{}, additionalCheck func(error)) (err error) {
```
Comment (tiancaiamao) [inline session/session_test.go:5275]: If you make `checkResult` a interface the switch (type) is unnecessary.
`checkError` do not actually need the arguments, so you can write

```
func (c checkSuccess) Check() {
    tk.MustExec(sql)
    tk.MustQuery("show warnings").Check(testkit.Rows())
    // inline the whole checkUpdatesAndDeletes code here
}

func (c checkError) Check() {
	err = tk.ExecToErr(sql)
	c.Assert(err, NotNil)
	expectedErr, _ := check.err.(*terror.Error)
    c.Assert(expectedErr.Equal(err), IsTrue)
     // inline the last line of checkUpdatesAndDeletes
}
```
`checkError` do not care much about how to check the result.
`checkSuccess`  have a `Check()` method and the reader would know how to check the result by reading that code.

File: session/session_test.go [hunk 3]
```diff
@@ -5149,3 +5150,185 @@ func (s *testSessionSuite) TestLocalTemporaryTableScan(c *C) {
...
+		_ = executeSql(sqlCase.sql, sqlCase.checkResult, sqlCase.additionalCheck)
+		tk.MustExec("delete from tmp1")
+
+		// update records in txn and rollback
+		insertRecords(idList)
+		tk.MustExec("begin")
+		_ = executeSql(sqlCase.sql, sqlCase.checkResult, sqlCase.additionalCheck)
+		tk.MustExec("rollback")
+		// rollback left records unmodified
+		checkUpdatesAndDeletes(nil, nil)
```
Comment (tiancaiamao) [inline session/session_test.go:5316]: You can define a `checkUnchanged()` to make `checkError.Check`  share the same logic.

**Comment by tiancaiamao** [comment]
Please fix the CI, rest looks good to me

**Review by tiancaiamao — APPROVED** [review]

**Review by AilinKid — APPROVED** [review]
LGTM

**Pull request merged**